Sources/SourceKit/SourceKitServer.swift

```diff
@@ -367,6 +371,31 @@ extension SourceKitServer {
     toolchainTextDocumentRequest(req, workspace: workspace, fallback: nil)
   }
 
+  func executeCommand(_ req: Request<ExecuteCommandRequest>, workspace: Workspace) {
+    guard let service = serviceFor(executeCommandRequest: req.params, workspace: workspace) else {
+      req.reply(nil)
+      return
+    }
+    let id = service.send(req.params, queue: DispatchQueue.global()) { result in
+      req.reply(result)
+    }
+    req.cancellationToken.addCancellationHandler { [weak service] in
+      service?.send(CancelRequest(id: id))
+    }
+  }
+
+  func serviceFor(executeCommandRequest req: ExecuteCommandRequest, workspace: Workspace) -> Connection? {
+    // FIXME: ExecuteCommand requests have no URL associated to it, but we need to determine the server
+    // that it gets sent to. There should be a better way to do this.
```

Comment:
can we pass url back in the `argument` field of `Commands` returned by `CodeActionResponse`?

Reply:
We can for Swift, but AFAIK we have no control over what clangd's commands will look like

Reply:
please correct me if i'm missing something, I assume parameters passed in the `CodeActionResponse` will be sent back to server when user picks the command. If that's the case, even for Clangd, server should still be the best place to generate those arguments, no?

Reply:
True, I believe we can intercept the response from clang and inject the url that called the code action itself into the arguments. It should be enough to avoid this logic, so we just need to confirm that doing so won't be a problem for clangd.

Reply:
True, I believe we can intercept the response from clang and inject the url that called the code action itself into the arguments. It should be enough to avoid this logic, so we just need to confirm that doing so won't be a problem for clangd.